Spigot-Server-Patches/0180-InventoryMoveItemEvent-improvements.patch

```diff
@@ -0,0 +1,165 @@
+-                        this.getWorld().getServer().getPluginManager().callEvent(event);
++                        // Paper start - InventoryMoveItemEvent improvements
++                        if (world.paperConfig.callMoveItemEvent) {
++                            boolean acceptsItem = willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);
++                            if (world.paperConfig.callMoveItemEventWithoutMove || acceptsItem) {
++                                if (world.paperConfig.callMoveEventCancelled && !acceptsItem) {
++                                    event.setCancelled(true);
++                                }
++                                this.getWorld().getServer().getPluginManager().callEvent(event);
++                            }
++                        }
++                        // this.getWorld().getServer().getPluginManager().callEvent(event);
++                        // Paper end
+                         if (event.isCancelled()) {
+                             this.setItem(i, itemstack);
+                             this.setCooldown(world.spigotConfig.hopperTransfer); // Spigot
+@@ -420,8 +431,18 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+             }
+ 
+             InventoryMoveItemEvent event = new InventoryMoveItemEvent(sourceInventory, oitemstack.clone(), ihopper.getOwner().getInventory(), false);
+-
+-            ihopper.getWorld().getServer().getPluginManager().callEvent(event);
++            // Paper start - InventoryMoveItemEvent improvements
++            if (ihopper.getWorld().paperConfig.callMoveItemEvent) {
++                boolean acceptsItem = willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);
++                if (ihopper.getWorld().paperConfig.callMoveItemEventWithoutMove || acceptsItem) {
++                    if (ihopper.getWorld().paperConfig.callMoveEventCancelled && !acceptsItem) {
++                        event.setCancelled(true);
++                    }
++                    ihopper.getWorld().getServer().getPluginManager().callEvent(event);
++                }
++            }
++            //ihopper.getWorld().getServer().getPluginManager().callEvent(event);
++            // Paper end
+             if (event.isCancelled()) {
+                 iinventory.setItem(i, itemstack1);
+ 
+@@ -452,6 +473,46 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+         return false;
+     }
+ 
++    // Paper start - method to check whether or not an inventory will accept the provided item. This should match Vanilla behaviour
++    // Vanilla method: addItem(IInventory, ItemsStack, EnumDirection)
++    static boolean willAcceptItem(IInventory iinventory, ItemStack itemstack, EnumDirection enumdirection) {
```

Comment:
This method only pulls the code that actually checks the inventory out of the addItem method which contains also the code to actually move the item. A better way might be the rewriting of the full addItem method instead of just pulling the code out, not sure how the policy is about such situation.